recipe-client-addon/content/shield-content-frame.js

```diff
@@ -0,0 +1,130 @@
+  frameGlobal, "AboutPages", "resource://shield-recipe-client-content/AboutPages.jsm",
+);
+
+const USE_OLD_PREF_ORGANIZATION_PREF = "browser.preferences.useOldOrganization";
+
+/**
+ * Handles incoming events from the parent process and about:studies.
+ * @implements nsIMessageListener
+ * @implements EventListener
+ */
+class ShieldFrameListener {
+  handleEvent(event) {
+    // Abort if the current page isn't about:studies.
+    if (!this.ensureTrustedOrigin()) {
+      return;
+    }
+
+    // We waited until after we received an event to register message listeners
+    // in order to save resources for tabs that don't ever load about:studies.
+    addMessageListener("Shield:ShuttingDown", this);
+    addMessageListener("Shield:ReceiveStudyList", this);
+
+    switch (event.detail.action) {
+      // Actions that require the parent process
+      case "GetRemoteValue:StudyList":
+        sendAsyncMessage("Shield:GetStudyList");
+        break;
+      case "RemoveStudy":
+        sendAsyncMessage("Shield:RemoveStudy", event.detail.data);
+        break;
+      // Actions that can be performed in the content process
+      case "GetRemoteValue:ShieldLearnMoreHref":
+        this.triggerPageCallback(
+          "ReceiveRemoteValue:ShieldLearnMoreHref",
+          frameGlobal.AboutPages.aboutStudies.getShieldLearnMoreHref()
+        );
+        break;
+      case "NavigateToDataPreferences":
+        this.navigateToDataPreferences();
+        break;
+    }
+  }
+
+  /**
+   * Check that the current webpage's origin is about:studies.
+   * @return {Boolean}
+   */
+  ensureTrustedOrigin() {
+    return content.document.documentURI.startsWith("about:studies");
+  }
+
+  /**
+   * Handle messages from the parent process.
+   * @param {Object} message
+   *   See the nsIMessageListener docs.
+   */
+  receiveMessage(message) {
+    switch (message.name) {
+      case "Shield:ReceiveStudyList":
+        this.triggerPageCallback("ReceiveRemoteValue:StudyList", message.data.studies);
+        break;
+      case "Shield:ShuttingDown":
+        this.onShutdown();
+        break;
+    }
+  }
+
+  /**
+   * Trigger an event to communicate with the unprivileged about: page.
+   * @param {String} type
+   * @param {Object} detail
+   */
+  triggerPageCallback(type, detail) {
+    // Do not communicate with untrusted pages.
+    if (!this.ensureTrustedOrigin()) {
+      return;
+    }
+
+    // Clone details and use the event class from the unprivileged context.
+    const event = new content.document.defaultView.CustomEvent(type, {
+      bubbles: true,
+      detail: Cu.cloneInto(detail, content.document.defaultView),
+    });
+    content.document.dispatchEvent(event);
+  }
+
+  onShutdown() {
+    removeMessageListener("Shield:SendStudyList", this);
+    removeMessageListener("Shield:ShuttingDown", this);
+    removeEventListener("Shield", this);
+  }
+
+  navigateToDataPreferences() {
+    if (Services.prefs.getBoolPref(USE_OLD_PREF_ORGANIZATION_PREF)) {
```

Comment:
Will this ever run against 54? If so, please add a fallback value of `true`, as it doesn't seem like the pref exists there.